### ISSUE TITLE
Feature to have export functionality for search results

### ISSUE THREAD

**Issue by akansha1102** [microsoft-search/pnp-modern-search#503]
**Is your feature request related to a problem? Please describe.**
We are using pnp modern search web part for document library search, we would like to request if there is a provision to export the search results for a specific keyword in the existing webpart or is there any future plan to incorporate the export functionality

**Describe the solution you'd like**
I have found a link while searching for this requirement:
https://social.technet.microsoft.com/Forums/en-US/820e088d-3185-45a8-9361-045cc449bab8/export-search-results-to-excel?forum=SP2019

![image](https://user-images.githubusercontent.com/64625884/94769417-9baf4280-03cf-11eb-8128-595a7a5daf21.png)

We would like to have a feature as shown in this picture above , not sure from where they have achieved it..as per their comment its an OOB solution, not sure if it's really is.

**Describe alternatives you've considered**

**Additional context**

**Comment by wobba** [comment]
No current plans, but you could create a web component as an extension to include this.

**Comment by quarterhorse** [comment]
It'd be nice to have an "Alert Me" in the results like the classic search result part has.

**Comment by wobba** [comment]
Not happening as the classic approach is sub-par imo. You can however implement something custom for this. @quarterhorse, it's also better to log new ideas as separate issues and not as comments on other topics. But no need to log this one 😉

**Comment by PzKfWg** [comment]
I also would like the opportunity to export to Excel results from the PNP Modern Search Web part. Maybe I could code it myself.

- Would it be possible for me to code it for the V3 version that I use the most? Or you have a policy that no further development should be done in V3?
- One challenge would be to know which properties to export to Excel. By default, a lot of default properties are pulled that are not likely to be useful in an Excel export. @akansha1102 , any thought? Or a section in the web part option to show the option or not and select the properties to export...?
- From a UI perspective, would it make sense to add the option beside the "Sort" control?
Thanks!

**Comment by wobba** [comment]
I think export of search results is an edge case. Do you export what is on page only, more results, which fields, how to format them?

If you come up with a good solution and PR for 3 and 4 we'll take it. Feel free to share how you envision setup to be before starting.

**Comment by PzKfWg** [comment]
Hi @wobba,
I started preliminary work on v3 to add the export feature. What I envision is a button beside the sort buttons:
![image](https://user-images.githubusercontent.com/24944003/117060300-c1e2eb00-acee-11eb-9cb0-5030ba9e50f9.png)
The button only appears if enabled on the property panel, third page:
![image](https://user-images.githubusercontent.com/24944003/117060410-e63ec780-acee-11eb-9f01-4936e1db4f8d.png)
There is then a button to select the fields to export, pretty much like when selecting the sorting fields:
![image](https://user-images.githubusercontent.com/24944003/117060559-07071d00-acef-11eb-84c2-e6df819ad36d.png)

What I now envision to do is that when the user clicks on the Excel icon, a CSV is built and available to download with all the search results (not only the current page) in a CSV format. I don't yet know if some limit could be hit if there are millions of results, so maybe I could cap the results to something manageable.

**Comment by PzKfWg** [comment]
I am trying to understand the SearchService to see indeed if I can request all the pages of the current request. I don't think that exporting only the current page makes sense. If I could put a progress bar and maybe an absolute ceil to the amount of results to be exported, it could be doable. The user that requested the feature wants to create an actionable Excel file based on the results, so PowerShell is out of question for them.

**Comment by wobba** [comment]
API has a limit for 500 items per page. And there is a limit for how far you can page back with a user query.

`Invoke-PnPSearchQuery -Query "my query" -All -RelevantResults | Export-Csv -Path .\file.csv` <- something like this will give you Excel. Super easy ;)

**Comment by PzKfWg** [comment]
Super easy for me for sure! But I don't know for you, but I have never seen an end-user open PowerShell, ever, in all of my clients :-P 
That limit for how far I can page with a user query, do you have the document at hand? This could help design some limit. I don't think it makes sense to export millions anyway.

**Comment by kasperbolarsen** [comment]
Perhaps it could be an option to do the heavy lifting in an Azure function,
and send the csv file as an email attachment?

fre. 7. maj 2021 20.35 skrev Francis ***@***.***>:

**Comment by PzKfWg** [comment]
I don't think it is possible to deploy an Azure artefact in a web part package...? The user might not even have an Azure subscription.

**Comment by wobba** [comment]
My point is, paging an API with many results is not suitable to do in javascript in the browser. It's better done in another container imo, allowing you to download the file when ready. Like what @kasperbolarsen suggests. If you cap to some limit javascript might be ok.

**Comment by sympmarc** [comment]
This came up with a client today. I understand all the reasons you give, @wobba. It's one of those "with great power..." situations.

**Comment by wobba** [comment]
I can imagine export to csv/pdf being possible of whatever is displayed in the result part. It should be export of on-page data, no additional API logic. API logic is certainly possible, but it quickly becomes very complicated. Also, paging/sorting the API based on anything but the internal id will not allow full paging. See https://docs.microsoft.com/en-us/sharepoint/dev/general-development/pagination-for-large-result-sets

**Comment by ellinge** [comment]
@PzKfWg any update here? Got a request from a client (through a colleague) here as well. Any branch to take a look at? Planning for supporting v4 as well?

@wobba so if anyone implements it as an export of current result set on client / javascript (no api call, thinking csv) it should be okey to include? Actually seems like a quite common request. Thinking in the terms of generating the csv in a string and put in the clipboard or similar if not possible to output as a "file" directly. Here some dated solution which perhaps works perfectly:
https://stackoverflow.com/a/14966131

**Comment by ellinge** [comment]
![image](https://user-images.githubusercontent.com/17863113/135541084-ebc8c8e1-1370-44ad-b229-cb616e801aae.png)
![image](https://user-images.githubusercontent.com/17863113/135541102-a4c09f30-842d-4486-bd60-4d3fad5fd964.png)

Have done some initial setup for configuration in v4. Starting off with DetailsList but trying to make it generic if found usable in other templates. Gonna have a break for now though.

**Comment by ellinge** [comment]
Got some inital thing going here for v4: https://github.com/ellinge/pnp-modern-search/tree/feature/export-to-csv

That above is only for the configuration though, will see when I find the time for "the rest".

**Comment by PzKfWg** [comment]
@ellinge , no update on my end. Thanks for looking at it! I like the solution that you refer to generate the CSV. 
I don't think handlebar would make much sense either.  I imagined that you could select the properties to export in the configuration panel and then just a csv/excel button in the results web part that gets the CSV (with an absolute limite to the number, like 100k or 1M depending on performance).

**Comment by ellinge** [comment]
@PzKfWg 
![](https://user-images.githubusercontent.com/17863113/135936532-446eb456-c88c-4a92-8e09-ee8ccc3df62a.gif)

I think more than 5k would be too slow. It takes some seconds to fetch. Guess one could make the calls async to speed it up a bit. But why would one want to export such a large result set either way? There’s a filter webpart for a reason 😄

**Comment by PzKfWg** [comment]
I love what I am seeing!!!
I did not test what would seem like a reasonable amount in a sync fashion. 
What sparked interest was that I deployed the modern search web part as a front-end to a "database" document library with ~30k items in it. There was a lot of cleanup necessary and the admin needed to export some lists with some filters, hence my need to be able to export a fair amount of results.
Anyway, I suspect that if we hard-limit to low, someone will for sure come up requesting a higher limit. I think the limit should be as high as bearable in a sync fashion. My 2-cents.

**Comment by ellinge** [comment]
@PzKfWg did a try now with Promise.all while exporting:
```
Processed '1, 2, 3, 4', items total fetched 2000
Processed '5, 6, 7, 8', items total fetched 4000
Processed '9, 10, 11, 12', items total fetched 6000
Processed '13, 14, 15, 16', items total fetched 8000
Processed '17, 18, 19, 20', items total fetched 9537
fetching: 16148.10595703125 ms
mapvalues: 14915.965087890625 ms
Processed 'TestWithFilename 2021-10-05.csv', items total exported 9537, has error column value: false, empty rows: 360
exporttocsv: 44.60693359375 ms
```
fetching = postquery against sharepoints api
mapvalues = fetching column values (including handlebars)
exporttocsv = push it to a blob for download

![exportascsvparallell](https://user-images.githubusercontent.com/17863113/136074377-92a97d7a-84b9-4258-8377-9a546de4d15d.gif)

**Comment by PzKfWg** [comment]
This looks to me like a very good balance, the 10k limit, between speed and usefulness. :-)
For the missing value, maybe call a full site reindex and wait a couple of hours to make sure that it isn't a simple index issue? I also have seen weird issues with the search index lately...

**Comment by kasperbolarsen** [comment]
Is there something I do to help you? I am not a developer, but I am pretty good at finding bugs and edge cases :-)

**Comment by JoeAyre** [comment]
Hi all, I notice the 'working on it' label was added back in October 2021. Just wondering if there is any update? The teaser gifs in previous comments look great, but I'm guessing the demand for it is not so great, and perhaps any work towards this has been put on hold?

**Comment by wobba** [comment]
I'll close this as the web part allow export on the last web part pane page 🙂
![image](https://user-images.githubusercontent.com/51104/186431354-3d0bc1c3-704c-4f1b-b828-d55dcfcbdd3b.png)

**Comment by JoeAyre** [comment]
He @wobba - apologies for perhaps appearing persistent on this but just wondering if this feature will ever make it into a release of the pnp-modern-search? If not, I may perhaps build the solution locally and see how I go but obviously if the feature is likely to added to the a future release I would prefer to wait. Any update would be greatly appreciated.
